File: wasm-wrapper/musig2.go
//go:build js && wasm

package main

// The musig2 namespace wraps the low-level BIP-327 MuSig2 functions of
// btcec/v2/schnorr/musig2 — deliberately the step-by-step primitives rather
// than the Session/Context API, so an educational UI can display every
// intermediate value of the two-round flow: aggregate keys, generate and
// aggregate nonces, partial-sign, combine.
//
// All functions use the BIP-327 convention of lexicographically sorted
// public keys, so callers can pass the key list in any order as long as
// they pass the same list everywhere.

import (
	"syscall/js"

	"github.com/btcsuite/btcd/btcec/v2"
	"github.com/btcsuite/btcd/btcec/v2/schnorr"
	"github.com/btcsuite/btcd/btcec/v2/schnorr/musig2"
)

// pubKeysFromArg parses a JS array of 33-byte compressed public keys.
func pubKeysFromArg(arg js.Value) ([]*btcec.PublicKey, map[string]any) {
	n := arg.Length()
	if n == 0 {
		return nil, errResult("at least one public key is required")
	}
	keys := make([]*btcec.PublicKey, n)
	for i := 0; i < n; i++ {
		raw, e := bytesFromArg(arg.Index(i))
		if e != nil {
			return nil, e
		}
		key, err := btcec.ParsePubKey(raw)
		if err != nil {
			return nil, errfResult("invalid public key %d: %s", i,
				err)
		}
		keys[i] = key
	}
	return keys, nil
}

// musig2AggregateKeys combines the signers' public keys into the single
// aggregated MuSig2 key.
// Calls Go: musig2.AggregateKeys() from btcec/schnorr/musig2.
func musig2AggregateKeys(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "pubKeys"); e != nil {
		return e
	}
	keys, e := pubKeysFromArg(args[0])
	if e != nil {
		return e
	}

	agg, _, _, err := musig2.AggregateKeys(keys, true)
	if err != nil {
		return errfResult("aggregate keys: %s", err)
	}

	compressed := agg.FinalKey.SerializeCompressed()
	return okResult(map[string]any{
		"combinedKey": bytesToJS(compressed),
		"xOnlyKey":    bytesToJS(schnorr.SerializePubKey(agg.FinalKey)),
		"parityOdd":   compressed[0] == 0x03,
	})
}

// musig2GenNonces generates the signer's secret/public nonce pair for one
// signing session. The public key is required (it is committed into the
// nonce derivation); the optional private key, combined key and message add
// further commitment entropy per BIP-327.
// Calls Go: musig2.GenNonces() from btcec/schnorr/musig2.
func musig2GenNonces(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "pubKey[, privKey, combinedKey, "+
		"msg]"); e != nil {

		return e
	}
	pubRaw, e := bytesFromArg(args[0])
	if e != nil {
		return e
	}
	pubKey, err := btcec.ParsePubKey(pubRaw)
	if err != nil {
		return errfResult("invalid public key: %s", err)
	}

	opts := []musig2.NonceGenOption{musig2.WithPublicKey(pubKey)}

	privRaw, e := optBytesFromArg(args, 1)
	if e != nil {
		return e
	}
	if privRaw != nil {
		priv, _ := btcec.PrivKeyFromBytes(privRaw)
		opts = append(opts, musig2.WithNonceSecretKeyAux(priv))
	}

	combinedRaw, e := optBytesFromArg(args, 2)
	if e != nil {
		return e
	}
	if combinedRaw != nil {
		combined, err := btcec.ParsePubKey(combinedRaw)
		if err != nil {
			return errfResult("invalid combined key: %s", err)
		}
		opts = append(opts, musig2.WithNonceCombinedKeyAux(combined))
	}

	msgRaw, e := optBytesFromArg(args, 3)
	if e != nil {
		return e
	}
	if msgRaw != nil {
		if len(msgRaw) != 32 {
			return errResult("msg must be 32 bytes")
		}
		var msg [32]byte
		copy(msg[:], msgRaw)
		opts = append(opts, musig2.WithNonceMessageAux(msg))
	}

	nonces, err := musig2.GenNonces(opts...)
	if err != nil {
		return errfResult("generate nonces: %s", err)
	}

	return okResult(map[string]any{
		"pubNonce": bytesToJS(nonces.PubNonce[:]),
		"secNonce": bytesToJS(nonces.SecNonce[:]),
	})
}

// musig2AggregateNonces combines all signers' public nonces into the single
// 66-byte combined nonce.
// Calls Go: musig2.AggregateNonces() from btcec/schnorr/musig2.
func musig2AggregateNonces(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "pubNonces"); e != nil {
		return e
	}

	n := args[0].Length()
	nonces := make([][musig2.PubNonceSize]byte, n)
	for i := 0; i < n; i++ {
		raw, e := bytesFromArg(args[0].Index(i))
		if e != nil {
			return e
		}
		if len(raw) != musig2.PubNonceSize {
			return errfResult("public nonce %d must be %d bytes",
				i, musig2.PubNonceSize)
		}
		copy(nonces[i][:], raw)
	}

	combined, err := musig2.AggregateNonces(nonces)
	if err != nil {
		return errfResult("aggregate nonces: %s", err)
	}
	return okResult(bytesToJS(combined[:]))
}

// musig2PartialSign creates one signer's partial signature. It returns the
// 32-byte partial signature s and the 33-byte final nonce R, which every
// signer computes identically and combineSigs needs.
// Calls Go: musig2.Sign() from btcec/schnorr/musig2.
func musig2PartialSign(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 5, "secNonce, privKey, combinedNonce, "+
		"pubKeys, msg"); e != nil {

		return e
	}

	secRaw, e := bytesFromArg(args[0])
	if e != nil {
		return e
	}
	if len(secRaw) != musig2.SecNonceSize {
		return errfResult("secNonce must be %d bytes",
			musig2.SecNonceSize)
	}
	var secNonce [musig2.SecNonceSize]byte
	copy(secNonce[:], secRaw)

	privRaw, e := bytesFromArg(args[1])
	if e != nil {
		return e
	}
	privKey, _ := btcec.PrivKeyFromBytes(privRaw)

	combinedRaw, e := bytesFromArg(args[2])
	if e != nil {
		return e
	}
	if len(combinedRaw) != musig2.PubNonceSize {
		return errfResult("combinedNonce must be %d bytes",
			musig2.PubNonceSize)
	}
	var combinedNonce [musig2.PubNonceSize]byte
	copy(combinedNonce[:], combinedRaw)

	keys, e := pubKeysFromArg(args[3])
	if e != nil {
		return e
	}

	msgRaw, e := bytesFromArg(args[4])
	if e != nil {
		return e
	}
	if len(msgRaw) != 32 {
		return errResult("msg must be 32 bytes")
	}
	var msg [32]byte
	copy(msg[:], msgRaw)

	partial, err := musig2.Sign(
		secNonce, privKey, combinedNonce, keys, msg,
		musig2.WithSortedKeys(),
	)
	if err != nil {
		return errfResult("partial sign: %s", err)
	}

	s := partial.S.Bytes()
	return okResult(map[string]any{
		"s": bytesToJS(s[:]),
		"r": bytesToJS(partial.R.SerializeCompressed()),
	})
}

// musig2CombineSigs combines all partial signatures (the 32-byte s values)
// with the final nonce R (returned by partialSign) into the final 64-byte
// BIP-340 Schnorr signature.
// Calls Go: musig2.CombineSigs() from btcec/schnorr/musig2.
func musig2CombineSigs(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 2, "finalNonce, partialSigs"); e != nil {
		return e
	}

	nonceRaw, e := bytesFromArg(args[0])
	if e != nil {
		return e
	}
	finalNonce, err := btcec.ParsePubKey(nonceRaw)
	if err != nil {
		return errfResult("invalid final nonce: %s", err)
	}

	n := args[1].Length()
	if n == 0 {
		return errResult("at least one partial signature is required")
	}
	partials := make([]*musig2.PartialSignature, n)
	for i := 0; i < n; i++ {
		raw, e := bytesFromArg(args[1].Index(i))
		if e != nil {
			return e
		}
		if len(raw) != 32 {
			return errfResult("partial signature %d must be 32 "+
				"bytes", i)
		}
		var s btcec.ModNScalar
		if overflow := s.SetByteSlice(raw); overflow {
			return errfResult("partial signature %d overflows "+
				"the curve order", i)
		}
		partials[i] = &musig2.PartialSignature{S: &s}
	}

	sig := musig2.CombineSigs(finalNonce, partials)
	return okResult(bytesToJS(sig.Serialize()))
}

File: src/musig2.ts
import { init, g, unwrap } from './init';
import type { Bytes } from './types';

/** The aggregated MuSig2 key from `musig2.aggregateKeys`. */
export interface Musig2AggregateKeysResult {
  /** 33-byte compressed aggregated key. */
  combinedKey: Uint8Array;
  /** 32-byte x-only aggregated key (for BIP-340 verification). */
  xOnlyKey: Uint8Array;
  /** Whether the aggregated key has an odd Y coordinate. */
  parityOdd: boolean;
}

/** One signer's nonce pair from `musig2.genNonces`. */
export interface Musig2Nonces {
  /** 66-byte public nonce (two compressed points), shared with cosigners. */
  pubNonce: Uint8Array;
  /** 97-byte secret nonce — single use, never shared. */
  secNonce: Uint8Array;
}

/** One signer's partial signature from `musig2.partialSign`. */
export interface Musig2PartialSignResult {
  /** 32-byte partial signature scalar. */
  s: Uint8Array;
  /** 33-byte final nonce R — identical for every signer; needed by
   *  `combineSigs`. */
  r: Uint8Array;
}

/** BIP-327 MuSig2 multi-signature utilities (two-round flow).
 *
 *  These wrap the low-level step-by-step functions rather than a session
 *  object, so every intermediate value is inspectable:
 *
 *  ```ts
 *  const agg = await musig2.aggregateKeys([pub1, pub2]);
 *  const n1 = await musig2.genNonces(pub1);           // each signer
 *  const n2 = await musig2.genNonces(pub2);
 *  const combined = await musig2.aggregateNonces([n1.pubNonce, n2.pubNonce]);
 *  const p1 = await musig2.partialSign(
 *    n1.secNonce, priv1, combined, [pub1, pub2], msgHash);
 *  const p2 = await musig2.partialSign(
 *    n2.secNonce, priv2, combined, [pub1, pub2], msgHash);
 *  const sig = await musig2.combineSigs(p1.r, [p1.s, p2.s]);
 *  // btcec.schnorrVerify(agg.xOnlyKey, msgHash, sig) === true
 *  ```
 *
 *  All functions use BIP-327's sorted-keys convention, so the public key
 *  list may be passed in any order — as long as it is the same list
 *  everywhere. */
export const musig2 = {
  /** Aggregate the signers' public keys into the single MuSig2 key.
   *  Calls Go: musig2.AggregateKeys() from btcd/btcec/schnorr/musig2. */
  async aggregateKeys(pubKeys: Bytes[]): Promise<Musig2AggregateKeysResult> {
    await init();
    return unwrap<Musig2AggregateKeysResult>(
      g().musig2.aggregateKeys(pubKeys),
    );
  },

  /** Generate one signer's secret/public nonce pair. The signer's public
   *  key is required; the optional private key, combined key and 32-byte
   *  message mix additional commitment entropy into the nonce derivation
   *  per BIP-327.
   *  Calls Go: musig2.GenNonces() from btcd/btcec/schnorr/musig2. */
  async genNonces(
    pubKey: Bytes,
    privKey?: Bytes,
    combinedKey?: Bytes,
    msg?: Bytes,
  ): Promise<Musig2Nonces> {
    await init();
    return unwrap<Musig2Nonces>(
      g().musig2.genNonces(pubKey, privKey, combinedKey, msg),
    );
  },

  /** Combine all signers' 66-byte public nonces into the combined nonce.
   *  Calls Go: musig2.AggregateNonces() from btcd/btcec/schnorr/musig2. */
  async aggregateNonces(pubNonces: Bytes[]): Promise<Uint8Array> {
    await init();
    return unwrap<Uint8Array>(g().musig2.aggregateNonces(pubNonces));
  },

  /** Create one signer's partial signature over the 32-byte message.
   *  Returns the partial signature `s` and the final nonce `r` (identical
   *  for all signers), which `combineSigs` needs.
   *  Calls Go: musig2.Sign() from btcd/btcec/schnorr/musig2. */
  async partialSign(
    secNonce: Bytes,
    privKey: Bytes,
    combinedNonce: Bytes,
    pubKeys: Bytes[],
    msg: Bytes,
  ): Promise<Musig2PartialSignResult> {
    await init();
    return unwrap<Musig2PartialSignResult>(
      g().musig2.partialSign(
        secNonce, privKey, combinedNonce, pubKeys, msg,
      ),
    );
  },

  /** Combine the partial signatures (32-byte `s` values) with the final
   *  nonce `r` into the final 64-byte BIP-340 Schnorr signature, valid
   *  under the aggregated x-only key.
   *  Calls Go: musig2.CombineSigs() from btcd/btcec/schnorr/musig2. */
  async combineSigs(
    finalNonce: Bytes,
    partialSigs: Bytes[],
  ): Promise<Uint8Array> {
    await init();
    return unwrap<Uint8Array>(
      g().musig2.combineSigs(finalNonce, partialSigs),
    );
  },
};
